### PR TITLE
Add `DcapEvidence` to `BlockMetadataContents`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
The `BlockMetadataContents` has been updated to accomodate the new
`DcapEvidence` that will be provided for attesting with the DCAP api.

Due to the change in `BlockMetadataContents` the block version has been
bumped.

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3554** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3554" target="_blank">Graphite</a> 
    * **PR #3555** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3555" target="_blank">Graphite</a> 
      * **PR #3556** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3556" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/mobilecoin/3556?utm_source=stack-comment).

File: api/src/convert/block_metadata.rs
Comment (nick-mobilecoin) [inline api/src/convert/block_metadata.rs:45]: While technically this captures `(true, true)` that shouldn't happen with a one of

File: blockchain/types/src/block_metadata.rs
Comment (nick-mobilecoin) [inline blockchain/types/src/block_metadata.rs:53]: I think we need to make sure this is the path we want to go.

Another option is to create a new  enum that contains `BlockMetadataContents` and `BlockMetadataContentsDcap` or similar and then during decode we key off of the block version for which variant to decode.

Comment (nick-mobilecoin) [inline blockchain/types/src/block_metadata.rs:53]: The use of `Option` here is similar to the comment on #3555 on questioning if we should be using prost up here at this level. This really isn't optional, there should always be one of the variants, but to work with prost this has to be an option.

File: mobilecoind/src/service.rs
Comment (nick-mobilecoin) [inline mobilecoind/src/service.rs:2612]: This used to be adding a block version 3 after block version 4 was already in the chain. Ledger will error out in these instances, https://github.com/mobilecoinfoundation/mobilecoin/blob/90560096c7cb306758b2d22bd4c5eb9891d20b66/ledger/db/src/ledger_db.rs#L695

File: api/src/convert/block_metadata.rs
Comment (eranrund) [inline api/src/convert/block_metadata.rs:32]: This shouldn't be necessary - I suggest looking at the generated code (search for `blockchain.rs` under `target`).
You'll find an enum called `BlockMetadataContents_oneof_attestation_evidence`, and you will be able to match against `src.attestation_evidence` which is an `Option<BlockMetadataContents_oneof_attestation_evidence>`

File: blockchain/types/src/block_metadata.rs
Comment (eranrund) [reply blockchain/types/src/block_metadata.rs:53]: I like this better, but it needs to be tested carefully. We must make sure that the digestible output of an existing BlockMetadataContents remains the same.

EDIT: saw the test for this, nice!

File: blockchain/types/src/block_metadata.rs
Comment (eranrund) [inline blockchain/types/src/block_metadata.rs:204]: Would be nice to test with something that is not all zeros, to represent a more real-world scenario.

**Review by eranrund — CHANGES_REQUESTED** [review]
I think this is the right approach. The conversion stuff can be cleaned up a bit (no need to the `match (src.has....)` stuff), and the test case can be improved to be more representative of real world (I am very cautious when touching anything that involves maintaining digestive-backwards-compatible behavior)

I fully agree we  should review this particular change very carefully, so more eyes on it would be good.

File: api/src/convert/block_metadata.rs
Comment (nick-mobilecoin) [reply api/src/convert/block_metadata.rs]: Logic has been updated to look at the `attestation_evidence` field of the `blockchain::BlockMetadataContents`.

I didn't realize they were public fields, thanks!

File: blockchain/types/src/block_metadata.rs
Comment (nick-mobilecoin) [reply blockchain/types/src/block_metadata.rs]: Test case has been updated to use `run_with_several_seeds` and `make_verification_report()`

**Comment by nick-mobilecoin** [comment]
> I think this is the right approach. The conversion stuff can be cleaned up a bit (no need to the `match (src.has....)` stuff), and the test case can be improved to be more representative of real world (I am very cautious when touching anything that involves maintaining digestive-backwards-compatible behavior)
> 
> I fully agree we should review this particular change very carefully, so more eyes on it would be good.

@eranrund I think I've addressed your comments, but didn't bump for re-review as I'm not sure if you blocked for the changes or to ensure more eyes happen here before merging

**Comment by eranrund** [comment]
> > I think this is the right approach. The conversion stuff can be cleaned up a bit (no need to the `match (src.has....)` stuff), and the test case can be improved to be more representative of real world (I am very cautious when touching anything that involves maintaining digestive-backwards-compatible behavior)
> > I fully agree we should review this particular change very carefully, so more eyes on it would be good.
> 
> @eranrund I think I've addressed your comments, but didn't bump for re-review as I'm not sure if you blocked for the changes or to ensure more eyes happen here before merging

I'll take a look. I blocked for the changes but more eyes would be good given the impact of this particular PR.

File: api/src/convert/block_metadata.rs
Comment (eranrund) [inline api/src/convert/block_metadata.rs:33]: Nit: You can merge the match statements and match on `Some(blockchain::BlockMetadataContents_oneof_attestation_evidence::dcap_evidence(evidence))` etc

File: blockchain/types/src/block_metadata.rs
Comment (eranrund) [reply blockchain/types/src/block_metadata.rs:53]: I'm not sure its similar since this is a oneof field...

**Review by eranrund — APPROVED** [review]

File: api/src/convert/block_metadata.rs
Comment (nick-mobilecoin) [reply api/src/convert/block_metadata.rs:33]: oh that would be better

File: blockchain/types/src/block_metadata.rs
Comment (nick-mobilecoin) [reply blockchain/types/src/block_metadata.rs:53]: true, it's a prost limitation in that there is no default enum variant when there isn't a simple enum, so to get a default, `Option<T>` was used.

File: api/src/convert/block_metadata.rs
Comment (nick-mobilecoin) [reply api/src/convert/block_metadata.rs]: updated, note I brought in `BlockMetadataContents_oneof_attestation_evidence` to use directly as it allowed keeping the match expression on one line for cargo fmt

**Comment by nick-mobilecoin** [comment]
closing in favor of #3577

**Pull request closed without merging**
